api_tests/lib/ledger_runner.ts

```diff
@@ -1,93 +1,187 @@
+
+                if (buffer.length > 0) {
+                    logFile.write("\n");
+                }
+            });
+
+            if (ledger === "bitcoin") {
+                bitcoin.init(await this.getBitcoinClientConfig());
                 this.blockTimers.bitcoin = global.setInterval(async () => {
                     await bitcoin.generate();
                 }, 1000);
             }
         }
     }
 
-    public stopLedgers() {
-        const names = Object.keys(this.runningLedgers);
-        if (names.length > 0) {
-            console.log("Stopping ledgers: " + names.join(", "));
+    public async stopLedgers() {
+        const ledgers = Object.entries(this.runningLedgers);
+
+        const promises = ledgers.map(async ([ledger, container]) => {
+            console.log(`Stopping ledger ${ledger}`);
+
+            clearInterval(this.blockTimers[ledger]);
+            await container.stop();
+            delete this.runningLedgers[ledger];
+        });
+
+        await Promise.all(promises);
+    }
+
+    public async getLedgerConfig(): Promise<LedgerConfig> {
+        return {
+            bitcoin: await this.getBitcoinClientConfig().catch(() => undefined),
+            ethereum: await this.getEthereumNodeConfig().catch(() => undefined),
+        };
+    }
+
+    private async getBitcoinClientConfig(): Promise<BitcoinNodeConfig> {
+        const container = this.runningLedgers.bitcoin;
+
+        if (container) {
+            const result = await container.exec([
+                "cat",
+                "/root/.bitcoin/regtest/.cookie",
+            ]);
+            const [, password] = result.output.split(":");
+
+            return {
+                host: container.getContainerIpAddress(),
+                rpcPort: container.getMappedPort(18443),
+                zmqPort: container.getMappedPort(28332),
+                username: "__cookie__",
+                password,
+            };
+        } else {
+            return Promise.reject("bitcoin not yet started");
+        }
+    }
+
+    private async getEthereumNodeConfig(): Promise<EthereumNodeConfig> {
+        const container = this.runningLedgers.ethereum;
 
-            Object.values(this.blockTimers).forEach(clearInterval);
+        if (container) {
+            const host = container.getContainerIpAddress();
+            const port = container.getMappedPort(8545);
 
-            execSync("docker-compose -f " + this.dockerComposeFile + " down");
+            return {
+                rpc_url: `http://${host}:${port}`,
+            };
+        } else {
+            return Promise.reject("ethereum not yet started");
         }
-        this.runningLedgers = {};
     }
+}
 
-    public getLedgersConfig(): any {
-        return this.ledgersConfig;
+/*
+ * For some weird reason, the log buffer contains weird prefixes
+ *
+ * This function removes those prefixes so that the logs can be printed to the file.
+ */
+function sanitizeBuffer(buffer: Buffer) {
```

Comment:
WTH... did you find out where this is coming from?

Reply:
Nope, no idea :sweat_smile: